packages/fuels-programs/src/retry.rs

```diff
@@ -0,0 +1,444 @@
+/// let interval_strategy = Backoff::Exponential(Duration::from_secs(1));
+///
+/// let retry_config = RetryConfig::new(max_attempts, interval_strategy);
+/// ```
+#[derive(Clone, Debug, Default)]
+pub struct RetryConfig {
+    pub max_attempts: usize,
+    pub interval: Backoff,
+}
+
+impl RetryConfig {
+    pub fn new(max_attempts: usize, interval: Backoff) -> Self {
+        RetryConfig {
+            max_attempts,
+            interval,
+        }
+    }
+}
+
+/// Retries an asynchronous action with customizable retry behavior.
+///
+/// This function takes an asynchronous action represented by a closure `action`.
+/// The action is executed repeatedly with backoff and retry logic based on the
+/// provided `retry_config` and the `should_retry` condition.
+///
+/// The `action` closure should return a `Future` that resolves to a `Result<T, K>`,
+/// where `T` represents the success type and `K` represents the error type.
+///
+/// # Parameters
+///
+/// - `action`: The asynchronous action to be retried.
+/// - `retry_config`: A reference to the retry configuration.
+/// - `should_retry`: A closure that determines whether to retry based on the result.
+///
+/// # Return
+///
+/// Returns `Ok(T)` if the action succeeds without requiring further retries.
+/// Returns `Err(K)` if the maximum number of attempts is reached and the action
+/// still fails. If a retryable error occurs during the attempts, the error will
+/// be returned if the `should_retry` condition allows further retries.
+///
+/// # Examples
+///
+/// ```rust
+/// use std::time::Duration;
+/// use tokio::time::sleep;
+/// use std::error::Error;
+/// use thiserror::Error;
+///
+/// use fuels_programs::retry::{Backoff, retry, RetryConfig};
+///
+/// #[derive(Debug, Error, Clone)]
+/// pub enum MyError {
+///     #[error("Network error: {0}")]
+///     NetworkError(String),
+/// }
+///
+/// async fn network_request() -> Result<(), MyError> {
+///     // Simulate network request here
+///     // ...
+///     // For demonstration purposes, always return an error
+///   Err(MyError::NetworkError("Network error".into()))
+/// }
+///
+/// fn main() {
+///     let retry_config = RetryConfig {
+///         max_attempts: 3,
+///         interval: Backoff::Linear(Duration::from_secs(1)),
+///     };
+///
+///     let should_retry = |result: &Result<(), MyError>| {
+///         // Retry if the error is retryable
+///         result.is_err()
+///     };
+///
+///     let result = retry(network_request, &retry_config, should_retry);
+/// }
+/// ```
+pub async fn retry<Fut, T, K, ShouldRetry>(
+    mut action: impl FnMut() -> Fut,
+    retry_config: &RetryConfig,
+    should_retry: ShouldRetry,
+) -> Result<T, K>
+where
+    T: Clone + Debug,
+    Fut: Future<Output = Result<T, K>>,
+    K: Clone + Error + 'static,
+    ShouldRetry: Fn(&Result<T, K>) -> bool,
+{
+    let mut last_err = None;
+    let max_attempts = retry_config.max_attempts;
+    for attempt in 0..max_attempts {
```

Comment:
If the count starts at 0 here, and the strategy is `Linear`, the first retry will occur with no delay. Is this expected of linear backoff?